src/lib/QueryParser.js

```diff
@@ -0,0 +1,459 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+foam.CLASS({
+  package: 'foam.parse',
+  name: 'QueryParser',
+  axioms: [foam.pattern.Singleton.create()],
+
+  // TODO(braden): Support KEYWORD predicates and queries on them.
+
+  requires: [
+    'foam.mlang.predicate.And',
+    'foam.mlang.predicate.ContainsIC',
+    'foam.mlang.predicate.Eq',
+    'foam.mlang.predicate.Gt',
+    'foam.mlang.predicate.Gte',
+    'foam.mlang.predicate.Has',
+    'foam.mlang.predicate.In',
+    'foam.mlang.predicate.InIC',
+    'foam.mlang.predicate.Lt',
+    'foam.mlang.predicate.Lte',
+    'foam.mlang.predicate.Not',
+    'foam.mlang.predicate.Or',
+    'foam.mlang.predicate.True',
+    'foam.parse.Alternate',
+    'foam.parse.ImperativeGrammar',
+    'foam.parse.LiteralIC',
+    'foam.parse.Parsers',
+    'foam.parse.StringPS'
+  ],
+
+  imports: [
```

Comment:
I'm not sure me should be imported, especially if this is a singleton/multiton.  Maybe it should be a parameter to the parse method, or should be part of the .create() arguments if we make this a multiton.